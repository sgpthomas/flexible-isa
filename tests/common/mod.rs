use std::{
    fmt::Display,
    path::{Path, PathBuf},
};

use flexible_isa::{cli, run, HalideExprOp, InstructionSelect};
use itertools::Itertools;

const BENCHMARKS_PATH: &str = "cache";
#[allow(unused)]
const BENCHMAKRS: &[&str] = &[
    "add",
    "average_pool",
    "blur3x3",
    "camera_pipe",
    "conv3x3a16",
    "conv3x3a32",
    "conv_nn",
    "depthwise_conv",
    "dilate3x3",
    // "elementwise",
    // "fully_connected",
    "gaussian3x3",
    "gaussian5x5",
    "gaussian7x7",
    "l2norm",
    "matmul",
    "max_pool",
    "mean",
    "median3x3",
    "mul",
    "sobel3x3",
    "softmax",
];

pub fn test_single_isa<S: AsRef<Path> + Display>(name: S) -> anyhow::Result<()> {
    let mut path = PathBuf::from(BENCHMARKS_PATH).join(&name);
    path.set_extension("stmt");

    let args: cli::Args = cli::Args::new(&[path]).learn(true);
    let isa = run(args)?;

    let instr_hist = InstructionSelect::from_recexpr(&isa.expressions);
    let hist: Vec<_> = instr_hist
        .iter()
        .filter_map(|(op, count)| {
            if let HalideExprOp::Instruction(i) = op {
                Some((i, count))
            } else {
                None
            }
        })
        // sort first by count, then by index
        .sorted_by_key(|(i, count)| (*count, *i))
        .map(|(i, count)| (i, count, isa.instructions[i].to_string()))
        .collect();

    insta::with_settings!({
        description => format!("generate isa for {name}"),
        omit_expression => true,
        snapshot_suffix => format!("{name}")
    }, {
        insta::assert_debug_snapshot!(hist)
    });

    Ok(())
}
