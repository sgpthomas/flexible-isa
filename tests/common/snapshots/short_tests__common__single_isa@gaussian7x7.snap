---
source: tests/common/mod.rs
description: generate isa for gaussian7x7
---
[
    (
        2,
        1,
        "(+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 4))) (cast<int32> ?x1))) (cast<int32> (neg (cast<int32> 3))))",
    ),
    (
        11,
        1,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 256))",
    ),
    (
        12,
        1,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 384))",
    ),
    (
        17,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> 1))",
    ),
    (
        21,
        1,
        "(+ (cast<int32> ?x0) (cast<int32> (neg (cast<int32> ?x1))))",
    ),
    (
        24,
        1,
        "(+ (cast<int32x128> (+ (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#10 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 6))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#11 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x2) (cast<int32> ?x3))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 15))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#12 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (- (cast<int32> ?x4) (cast<int32> ?#13))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 20))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#14 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (- (cast<int32> ?x4) (cast<int32> ?#15))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 15))))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#16 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x5) (cast<int32> ?x6))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 6))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#17 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x7) (cast<int32> ?x8))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        38,
        1,
        "(+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#13 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 6))))) (cast<int32x128> (+ (cast<int32x128> (get ?#14 (cast<int32x128> (call[ramp] (cast<int32> ?x2) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#15 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x3))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 15))))) (cast<int32x128> (+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#16 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x4))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 20))))) (cast<int32x128> (+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#17 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 15))))) (cast<int32x128> (+ (cast<int32x128> (get ?#18 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x6))) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#19 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x7))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 6))))))))))))))))",
    ),
    (
        56,
        1,
        "(- (cast<int32> (* (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        67,
        1,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        68,
        1,
        "(* (cast<int32> ?x0) (cast<int32> 2))",
    ),
    (
        69,
        1,
        "(* (cast<int32> ?x0) (cast<int32> 3))",
    ),
    (
        79,
        1,
        "(== ?#0 ?#1)",
    ),
    (
        82,
        1,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> (* (cast<int32> ?#2) (cast<int32> 4))) (cast<int32> ?x0))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        9,
        2,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 32))) (cast<int32> ?x1))",
    ),
    (
        10,
        2,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> 128))",
    ),
    (
        15,
        2,
        "(+ (cast<int32> (* (cast<int32> ?#1) (cast<int32> 2))) (cast<int32> ?x0))",
    ),
    (
        16,
        2,
        "(+ (cast<int32> (* (cast<int32> ?#1) (cast<int32> 3))) (cast<int32> ?x0))",
    ),
    (
        58,
        2,
        "(- (cast<int32> (* (cast<int32> ?#1) (cast<int32> (neg (cast<int32> ?x0))))) (cast<int32> ?#2))",
    ),
    (
        63,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 1))) (cast<int32> ?x1))",
    ),
    (
        64,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 2))) (cast<int32> ?x1))",
    ),
    (
        65,
        2,
        "(* (cast<int32> (+ (cast<int32> ?x0) (cast<int32> 3))) (cast<int32> ?x1))",
    ),
    (
        73,
        2,
        "(* (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        23,
        3,
        "(+ (cast<int32x128> (+ (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#6 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#7) (cast<int32> ?x0))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 6))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#8 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#9) (cast<int32> ?x0))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 15))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#10 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?#11))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 20))))) (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#12 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (- (cast<int32> ?x0) (cast<int32> ?#13))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 15))))))))) (cast<int32x128> (call[widening_add] (cast<int16x128> (call[widening_mul] (cast<uint8x128> (get ?#14 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#15) (cast<int32> ?x0))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))) (cast<int8x128> (call[x128] (cast<int8> 6))))) (cast<int16x128> (call[int16x128] (cast<uint8x128> (get ?#16 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?#17) (cast<int32> ?x0))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))))))))",
    ),
    (
        39,
        3,
        "(+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#14 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 6))))) (cast<int32x128> (+ (cast<int32x128> (get ?#15 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x2))) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#16 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x3))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 15))))) (cast<int32x128> (+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#17 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x4))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 20))))) (cast<int32x128> (+ (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#18 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x5))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 15))))) (cast<int32x128> (+ (cast<int32x128> (get ?#19 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x6))) (cast<int32> 1) (cast<int32> 128))))) (cast<int32x128> (call[widen_right_mul] (cast<int32x128> (get ?#20 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x7))) (cast<int32> 1) (cast<int32> 128))))) (cast<int16x128> (call[x128] (cast<int16> 6))))))))))))))))",
    ),
    (
        59,
        3,
        "(- (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        72,
        3,
        "(* (cast<int32> ?x0) (cast<int32> 128))",
    ),
    (
        76,
        3,
        "(/ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?x1))) (cast<int32> ?x2))",
    ),
    (
        83,
        3,
        "(get ?#1 (cast<int32x128> (call[ramp] (cast<int32> (+ (cast<int32> (+ (cast<int32> ?x0) (cast<int32> ?#2))) (cast<int32> (neg (cast<int32> 3))))) (cast<int32> 1) (cast<int32> 128))))",
    ),
    (
        22,
        5,
        "(+ (cast<int32> ?x0) (cast<int32> ?x1))",
    ),
    (
        13,
        6,
        "(+ (cast<int32> (* (cast<int32> ?x0) (cast<int32> 128))) (cast<int32> ?x1))",
    ),
    (
        70,
        8,
        "(* (cast<int32> ?x0) (cast<int32> 4))",
    ),
]
